File: codegen/objc/objc-entities/src/main/groovy/com/stanfy/helium/handler/codegen/objectivec/entity/file/ObjCClassImplementation.java
package com.stanfy.helium.handler.codegen.objectivec.entity.file;

import java.util.ArrayList;

/**
 * Created by ptaykalo on 8/17/14.
 * Simple block that know how to serialize ObjC Class Implementation
 */
public class ObjCClassImplementation implements ObjCSourcePart {

  /*
  Class Name
   */
  private String className;
  private ArrayList<ObjCSourcePart> importSourceParts =  new ArrayList<ObjCSourcePart>();
  private ArrayList<ObjCSourcePart> bodySourceParts =  new ArrayList<ObjCSourcePart>();

  public ObjCClassImplementation(final String className) {
    this.className = className;
  }

  public String getClassName() {
    return className;
  }

  /**
  Adds specified source part to the top part (before @implementation)
   */
  public void addImportSourcePart(final ObjCSourcePart sourcePart) {
    importSourceParts.add(sourcePart);
  }
  /**
   Adds specified source part to the central part (inside @implementation)
   */
  public void addBodySourcePart(final ObjCSourcePart sourcePart) {
    bodySourceParts.add(sourcePart);
  }

  @Override
  public String asString() {
    // TODO use some templates
    StringBuilder bld = new StringBuilder();
    bld.append("#import \"").append(className).append(".h\"\n");
    for (ObjCSourcePart sourcePart : importSourceParts) {
      bld.append(sourcePart.asString()).append("\n");
    }
    bld.append("@implementation ").append(className).append("\n");
    for (ObjCSourcePart sourcePart : bodySourceParts) {
      bld.append(sourcePart.asString()).append("\n");
    }
    bld.append("@end");
    return bld.toString();

  }
}

File: codegen/objc/objc-entities/src/main/kotlin/com/stanfy/helium/handler/codegen/objectivec/entity/classtree/ObjCClassImplementation.kt
package com.stanfy.helium.handler.codegen.objectivec.entity.classtree

import com.stanfy.helium.handler.codegen.objectivec.entity.filetree.ObjCImportPart
import com.stanfy.helium.handler.codegen.objectivec.entity.filetree.ObjCSourcePart
import java.util.ArrayList

/**
 * Created by ptaykalo on 8/17/14.
 * Simple block that know how to serialize ObjC Class Implementation
 */
class ObjCClassImplementation(val filename: String) : ObjCSourcePart {

  // TODO : Dependencies
  private val importSourceParts = hashSetOf<ObjCImportPart>()
  private val bodySourceParts = arrayListOf<ObjCSourcePart>()

  /**
  Adds specified source part to the top part (before @implementation)
   */
  fun addImportSourcePart(sourcePart: ObjCImportPart) {
    importSourceParts.add(sourcePart)
  }

  fun importClassWithName(className:String) {
    this.addImportSourcePart(ObjCImportPart(className))
  }
  /**
  Adds specified source part to the central part (inside @implementation)
   */
  fun addBodySourcePart(sourcePart: ObjCSourcePart) {
    bodySourceParts.add(sourcePart)
  }

  /**
   * Adds method implementation part to the
   */
  fun addMethod(method:ObjCMethod) {
    addBodySourcePart(ObjCMethodImplementationSourcePart(method))
  }


  override fun asString(): String {
    val bld = StringBuilder()
    bld.append("#import \"").append(filename).append(".h\"\n")
    for (sourcePart in importSourceParts) {
      bld.append(sourcePart.asString()).append("\n")
    }
    bld.append("@implementation ").append(filename).append("\n")
    for (sourcePart in bodySourceParts) {
      bld.append(sourcePart.asString()).append("\n")
    }
    bld.append("@end")
    return bld.toString()

  }
}
